### PR TITLE
Flatpak manifest

### PR DESCRIPTION
**Pull request by p1u3o**
This is an initial Flatpak manifest for Cawbird based on the older Corebird manifest. It builds and works.

I'm opening this here for feedback.

### CODE REVIEW

**Comment by vrutkovs** [comment]
I've been created https://github.com/vrutkovs/flatpak-cawbird/blob/master/uk.co.ibboard.Cawbird.json and tested it. This one looks good, except it needs `enchant`. Also it would be nice to use Gnome 3.34

**Comment by p1u3o** [comment]
Thanks @vrutkovs 

I've updated it to use the nightly Gnome SDK. The Flathub manifest can use 3.34 once https://github.com/IBBoard/cawbird/commit/7b88f8856d108b9555ba7b855c7daed7b9e745ca is in a release.

File: uk.co.ibboard.cawbird.json [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      "sources":[
+        {
+          "type":"archive",
+          "url":"https://gstreamer.freedesktop.org/src/gst-libav/gst-libav-1.16.1.tar.xz",
+          "sha256":"e8a5748ae9a4a7be9696512182ea9ffa6efe0be9b7976916548e9d4381ca61c4"
+        }
+      ]
+    },
+    {
+      "name":"enchant",
```
Comment (TingPing) [inline uk.co.ibboard.cawbird.json]: gspell supports enchant 2 which is in the runtime last i remember.

File: uk.co.ibboard.cawbird.json [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  "runtime-version":"master",
+  "sdk":"org.gnome.Sdk",
+  "command":"cawbird",
+  "desktop-file-name-prefix": "(Development) ",
+  "rename-icon":"cawbird",
+  "copy-icon":true,
+  "finish-args":[
+    "--share=ipc",
+    "--share=network",
+    "--socket=x11",
```
Comment (TingPing) [inline uk.co.ibboard.cawbird.json]:
```suggestion
    "--socket=fallback-x11",
```

File: uk.co.ibboard.cawbird.json [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    "--socket=wayland",
+    "--socket=pulseaudio",
+    "--filesystem=home"
```
Comment (TingPing) [inline uk.co.ibboard.cawbird.json]: This shouldn't be needed. I've not looked to see if this project doesn't use GtkFileChooserNative but I can't imagine why it needs full home access.

File: uk.co.ibboard.cawbird.json [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    "/lib/*.la",
+    "/lib/libvala*",
+    "/lib/vala-0.34",
+    "/share/gtk-doc",
+    "/share/man",
+    "/share/vala"
+  ],
+  "modules":[
+    {
+      "name":"gst-libav",
```
Comment (TingPing) [inline uk.co.ibboard.cawbird.json]: This also probably isn't needed as its in the runtime.

File: uk.co.ibboard.cawbird.json [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  "finish-args":[
+    "--share=ipc",
+    "--share=network",
+    "--socket=x11",
+    "--socket=wayland",
+    "--socket=pulseaudio",
+    "--filesystem=home"
+  ],
+  "build-options":{
+    "cflags":"-O2 -g"
```
Comment (TingPing) [inline uk.co.ibboard.cawbird.json]: This can be removed its the default CFLAGS already.

File: uk.co.ibboard.cawbird.json (hunk 4 above)
Comment (p1u3o) [reply uk.co.ibboard.cawbird.json]: with h264 support?

File: uk.co.ibboard.cawbird.json (hunk 1 above)
Comment (TingPing) [reply uk.co.ibboard.cawbird.json]: Actuall its not in the runtime, but you should probably still upgrade to enchant 2.x.

File: uk.co.ibboard.cawbird.json (hunk 4 above)
Comment (p1u3o) [reply uk.co.ibboard.cawbird.json]: nm seems to work

File: uk.co.ibboard.cawbird.json (hunk 3 above)
Comment (p1u3o) [reply uk.co.ibboard.cawbird.json]: Done

**Comment by p1u3o** [comment]
Thanks @TingPing 

Let me know if there's anything else you think should be changed. I think org.freedesktop.Platform.openh264 is required for videos to play correctly (at least from my testing), so does that need to be explicitly mentioned as an extension?

File: uk.co.ibboard.cawbird.json [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    "/share/aclocal",
+    "/lib/*.la",
+    "/lib/libvala*",
+    "/lib/vala-0.34",
+    "/share/gtk-doc",
+    "/share/man",
+    "/share/vala"
+  ],
+  "add-extensions":{
+    "org.freedesktop.Platform.ffmpeg-full":{
```
Comment (TingPing) [inline uk.co.ibboard.cawbird.json]: You don't have to add this. Since its an extension point of the runtime it is always installed automatically anyway.

Comment (p1u3o) [reply uk.co.ibboard.cawbird.json]: I'm not sure why, but without this videos freeze after 10 or so seconds.

**Comment by TingPing** [comment]
> Let me know if there's anything else you think should be changed. I think org.freedesktop.Platform.openh264 is required for videos to play correctly (at least from my testing), so does that need to be explicitly mentioned as an extension?

Extension points are always optional so you have two options:

- Accept users may forcefully remove that extension or their flatpak is too old to install it (the latter is the only concern really).
- Build `ffmpeg` yourself with h264 enabled (effectively what `gst-libav` was doing before, but you only wanted `ffmpeg` not a gstreamer plugin.

I don't think either approach is wrong but the former requires less maintainance and worst case you can document very old `flatpak` versions will miss it by default.

File: uk.co.ibboard.cawbird.json (hunk 6 above)
Comment (TingPing) [reply uk.co.ibboard.cawbird.json]: Hmm... I wonder if that is openh264 being used then.

It is harmless to leave I suppose.

Comment (p1u3o) [reply uk.co.ibboard.cawbird.json]: It can be re-visited later, someone might have a better idea of why gstreamer freezes after 10s without it.

**Comment by p1u3o** [comment]
I think the first option is going to be okay for now with ffmpeg-full. The streaming is still quite glitchy but it doesn't freeze outright now.

**Update**

So I had a look at why Totem didn't have the same issues, and it seems building the gst plugins with OpenH264 disabled is enough to fix it completely, videos play near-perfect now.

**Comment by p1u3o** [comment]
This is probably good to merge now. I can't find any more issues.

**Comment by TingPing** [comment]
> So I had a look at why Totem didn't have the same issues, and it seems building the gst plugins with OpenH264 disabled is enough to fix it completely, videos play near-perfect now.

Hmm. We should probably report this against the runtime to maybe patch gstreamer to prefer ffmpeg over openh264.

Also you probably don't need to build all of -bad and -ugly too? Twitter surely uses a single codec.

**Comment by p1u3o** [comment]
> Also you probably don't need to build all of -bad and -ugly too? Twitter surely uses a single codec.

I think h264 (and maybe hevc too for future proofing?) should cover it, I'll prepare a MR to remove everything else.

ugly can probably be removed.

**Comment by IBBoard** [comment]
For reference, the [spec files](https://build.opensuse.org/package/view_file/home:IBBoard:cawbird/cawbird/cawbird.spec) that I build with just reference `gstreamer-plugins-libav` or equivalent.

**Comment by TingPing** [comment]
https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/903

**Comment by TingPing** [comment]
It may make sense for cawbird to create a custom gstreamer pipeline explicitly using gst-libav. This isn't a Flatpak specific issue that gstreamer prefers openh264 it just happens to be the first place to deploy this and it will soon be broken in Fedora and more to come.

**Comment by TingPing** [comment]
Ok so correction: The way you opt-in to ffmpeg-full is correct and the gstreamer priorities are correct. So you can just remove all gstreamer packages, re-add the extension, and everything should be correct. Sorry about the confusion.

**Comment by p1u3o** [comment]
Problem is, whilst enabling the ffmpeg extension fixed the streams freezing, it brought in a lot of graphical glitches that building the plugins ourselves doesn't seem to do.

**Comment by TingPing** [comment]
Hmm, can you run this command: `flatpak run --env=GST_DEBUG=4 --command=gst-launch-1.0 uk.co.ibboard.cawbird playbin uri="file:///path/to/mp4"`

**Comment by p1u3o** [comment]
building our own gst pugins: https://paste.fedoraproject.org/paste/-NPE1cg4DB8L7XPnvuK5QA
using the ffmpeg-full extension: https://paste.fedoraproject.org/paste/jtbaLVwe-UqeGWgmzXk7Zg

I start to see differences on line 382.

gst
```
GST_EVENT gstevent.c:820:gst_event_new_caps: creating caps event audio/mpeg, mpegversion=(int)4, framed=(boolean)true, stream-format=(string)raw, level=(string)2, base-profile=(string)lc, profile=(string)lc, codec_data=(buffer)1190, rate=(int)48000, channels=(int)2
```

ffmpeg
```
GST_EVENT gstevent.c:820:gst_event_new_caps: creating caps event video/x-h264, stream-format=(string)avc, alignment=(string)au, level=(string)3.1, profile=(string)high, codec_data=(buffer)0164001fffe100186764001facb605a0b742000003000200000300651e3065c001000568eac272c0, width=(int)720, height=(int)720, framerate=(fraction)25/1, pixel-aspect-ratio=(fraction)1/1, interlace-mode=(string)progressive, chroma-format=(string)4:2:0, bit-depth-luma=(uint)8, bit-depth-chroma=(uint)8, parsed=(boolean)true
	0:00:00.150455657     2 0x7f7e7c083b20 INFO                GST_PADS gstpad.c:2377:gst_pad_link_prepare: trying to link qtdemux0:audio_0 and decodepad2:proxypad3		
```

**Comment by TingPing** [comment]
Ok so the expected decoders are being used. If ffmpeg is causing problems all I could say is maybe try building an older version to see if its a regression.

I wonder whats so broken about Twitter videos that 2 decoders can't even handle them...

**Comment by IBBoard** [comment]
I've had issues with [Corebird builds and GStreamer dependencies](https://twitter.com/IBBoard/status/1160908676405874689). That was more of an "all or nothing" thing, though - either it failed or it worked. It was about the difference between gstreamer-plugins-bad and libav (both of which claimed to support h.264)

**Comment by p1u3o** [comment]
> Ok so the expected decoders are being used. If ffmpeg is causing problems all I could say is maybe try building an older version to see if its a regression.

I wonder if it's a regression in 1.16.1 in the GNOME 3.34 runtime (we'll building 1.16). I'll have a test.

**Comment by p1u3o** [comment]
Let's continue this in https://github.com/IBBoard/cawbird/pull/42

**Pull request merged**